appinventor/components/src/com/google/appinventor/components/runtime/PieChartDataModel.java

```diff
@@ -0,0 +1,221 @@
+
+      // Add the Legend Entry both to the local Legend Entries List and
+      // to the Legend of the view itself.
+      legendEntries.add(legendEntry);
+      view.addLegendEntry(legendEntry);
+    }
+  }
+
+  @Override
+  public void removeEntryFromTuple(YailList tuple) {
+    // Construct an entry from the specified tuple
+    Entry entry = getEntryFromTuple(tuple);
+
+    if (entry != null) {
+      // Get the index of the entry
+      int index = findEntryIndex(entry);
+
+      // Entry exists; remove it
+      if (index >= 0) {
+        getDataset().removeEntry(index);
+        LegendEntry removedEntry = legendEntries.remove(index);
+        view.removeLegendEntry(removedEntry);
+      }
+    }
+  }
+
+  @Override
+  public Entry getEntryFromTuple(YailList tuple) {
+    try {
+      // Tuple is expected to have at least 2 entries.
+      // The first entry is assumed to be the x value, and
+      // the second is assumed to be the y value.
+      String xValue = tuple.getString(0);
+      String yValue = tuple.getString(1);
+
+      try {
+        // Attempt to parse the y value String representation
+        float y = Float.parseFloat(yValue);
+
+        // The y value is the first argument,
+        // the x value is the second (label)
+        return new PieEntry(y, xValue);
+      } catch (NumberFormatException e) {
+        // Nothing happens: Do not add entry on NumberFormatException
+      }
+    } catch (Exception e) {
+      // 2-tuples are invalid when null entries are present, or if
+      // the number of entries is not sufficient to form a pair.
+      // TODO: Show toast error notification
+    }
+
+    return null;
+  }
+
+  @Override
+  public YailList getTupleFromEntry(Entry entry) {
+    // Cast Entry to PieEntry (safe cast)
+    PieEntry pieEntry = (PieEntry) entry;
+
+    // Create a list with the X and Y values of the entry, and
+    // convert the generic List to a YailList
+    List tupleEntries = Arrays.asList(pieEntry.getLabel(), pieEntry.getY());
+    return YailList.makeList(tupleEntries);
+  }
+
+  @Override
+  protected void setDefaultStylingProperties() {
+    // Set spacing between each slice
+    dataset.setSliceSpace(3);
+  }
+
+  @Override
+  protected YailList getDefaultValues(int size) {
+    // Default values for PieChartBaseDataModel should be
+    // integers from 0 to N (0, 1, 2, ...)
+    // TODO: This could be updated in the future to return Strings
+    // TODO: such as "Entry 1", "Entry 2", ... for x and 1,2,3,...,N for y
+    ArrayList<Integer> defaultValues = new ArrayList<>();
+
+    for (int i = 0; i < size; ++i) {
+      defaultValues.add(i);
+    }
+
+    return YailList.makeList(defaultValues);
+  }
+
+  /**
+   * Sets the colors of the Data Series from the specified
+   * colors YailList.
+   *
+   * Invalid (non-color) entries are simply ignored.
+   * @param colors YailList of entries expected to contain colors (integers)
+   */
+  public void setColors(YailList colors) {
```

Comment:
Test cases:
- [ ] 1 color, multi entries
- [ ] N colors, multi entries (all have it's own color)
- [ ] N colors, multi entries (repeated colors)
- [ ] Invalid entries in between